concent_api/middleman/middleman_server.py

```diff
@@ -120,52 +124,86 @@ def _close_middleman(self) -> None:
+            await request_producer_task
+            await asyncio.sleep(PROCESSING_TIMEOUT)
+            await response_queue.join()
+            response_consumer_task.cancel()
+
         except Exception as exception:  # pylint: disable=broad-except
             crash_logger.error(
                 f"Exception occurred: {exception}, Traceback: {traceback.format_exc()}"
             )
             raise
+        finally:
+            for task in tasks:
+                task.cancel()
+            while response_queue.qsize() > 0:
+                item = await response_queue.get()  # type: ResponseQueueItem
+                logger.info(
+                    f"Dropped message: {type(item.message)}, "
+                    f"Concent request ID: {item.concent_request_id}, "
+                    f"connection ID: {self._connection_id}"
+                )
+            removed_queue = self._response_queue_pool.pop(self._connection_id, None)
+            if removed_queue is None:
+                logger.warning(f"Response queue for connection ID: {self._connection_id} has been already removed")
+            else:
+                logger.info(f"Removing response queue for connection ID: {self._connection_id}.")
 
     async def _handle_service_connection(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
         if self._is_signing_service_connection_active:
             writer.close()
         else:
             self._is_signing_service_connection_active = True
+            tasks = []
             try:
-                frame = await handle_frame_receive_async(reader, settings.CONCENT_PUBLIC_KEY)
-                remote_address = writer.get_extra_info('peername')
-                print("Received %r from %r" % (frame, remote_address))
-                await self._respond_to_user(frame, writer)
-            except (BrokenEscapingInFrameMiddlemanProtocolError, asyncio.LimitOverrunError, MiddlemanProtocolError) as exception:
-                await self._send_immediate_error(writer, exception)
+                request_consumer_task = self._loop.create_task(
+                    request_consumer(
+                        self._request_queue,
+                        self._response_queue_pool,
+                        self._message_tracker,
+                        writer
+                    )
+                )
+                response_producer_task = self._loop.create_task(
+                    response_producer(
+                        self._response_queue_pool,
+                        reader,
+                        self._message_tracker
+                    )
+                )
+                futures = [request_consumer_task, response_producer_task]
+                tasks = futures[:]
+                done, pending = await asyncio.wait(futures, return_when=asyncio.FIRST_COMPLETED)
+                for future in pending:
+                    future.cancel()
+                for future in done:
```

Comment:
Isn't `done` a single element ?

Reply:
no, it is a set of `Future`s (https://docs.python.org/3/library/asyncio-task.html#asyncio.wait); I will rename it to be less ambiguous;
done